VACMS-8360 Homepage hero component functionality for random images

allisonlu opened the pull request:
**DO NOT MERGE** - PR is in progress, waiting for PO approval

## Description
Functionality for Public Websites's [issue #8360](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8360) - randomizes images in hero component. Takes an array of Veteran portraits (provided by Ryan Thurlwell), randomizes it, places a background-color depending on its place, injects into hero. 

## Original issue(s)
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8360


## Testing done
- [x] locally
- [ ] user testing pending

## Screenshots
### Mobile
<img width="363" alt="Screen Shot 2022-05-06 at 4 41 34 PM" src="https://user-images.githubusercontent.com/8542413/167213800-27420dd1-5446-4f0c-afda-b4ed07c7b79b.png">

### Tablet
<img width="797" alt="Screen Shot 2022-05-06 at 4 41 48 PM" src="https://user-images.githubusercontent.com/8542413/167213789-6dd439e7-360a-4acc-bd6f-26738297ccf1.png">

### Desktop
<img width="1188" alt="Screen Shot 2022-05-06 at 4 42 02 PM" src="https://user-images.githubusercontent.com/8542413/167213779-3fbbceb6-9bc3-48fc-9a5b-de3f63a6490f.png">

## Acceptance criteria
- [ ] Different random order of veterans on page load
- [ ] Responsive (stacks on mobile)

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

allisonlu committed:
VACMS-8360 set up architecture for home page hero React widget

allisonlu committed:
VACMS-8360
- created a component for the Veteran images in the hero with logic that randomizes the order on pageload

allisonlu committed:
VACMS-8360:
- created scss stylesheet for homepage hero
- added some styles for positioning

allisonlu committed:
VACMS-8360:
- background a gradient background color that goes off the grid
- added background colors to each image

allisonlu committed:
VACMS-8360:
- added documentation to stylesheet
- added gradient background color that goes off the grid
- centered images better in container

allisonlu committed:
VACMS-8360:
- added styles for button

allisonlu committed:
VACMS-8360 changed height to min-height for Safari

allisonlu committed:
Merge branch 'master' into VACMS-8360-home-hero-random-image

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+  return tempArray.slice(0, 3);
+};
+
+const HeroRandom = () => {
```
I'm not sure an `id` property is necessary here. As such, I think this array can be altered to be an array of strings rather than objects.

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+
+  // randomize order of images on pageload
+  useEffect(() => {
+    setRandomized(randomizeOrder(imagePaths));
```
See GHA/Linting error below. Update the second parameter to the function to be `[imagePaths]` rather than `[]`.

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+    { id: 8, name: 'singleton' },
+  ];
+
+  const gradientValues = ['#105f9f', '#144e84', '#143e6a'];
```
I think this should likely be done in scss.  This brings to light the idea that we'll likely need to add these colors to the approved VA design system.  This might take some extra work/governance.

allisonlu replied:
As discussed in scrum 4/21, we will keep it simple for this iteration, then refine later.

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+      <div className="vads-l-col" key={image.id}>
+        <img
+          className="homepage-hero__image"
+          style={{ backgroundColor: gradientValues[index] }}
```
See comment about this being done in scss/design system.

allisonlu replied:
As discussed in scrum 4/21, we will keep it simple for this iteration, then refine later.

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+
+  return randomized.map((image, index) => {
+    return (
+      <div className="vads-l-col" key={image.id}>
```
The key can be the image name, since the names are unique.  This id approach is nice, but, as noted above, is likely unnecessary and the array can be simplified.

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+    setRandomized(randomizeOrder(imagePaths));
+  }, []);
+
+  return randomized.map((image, index) => {
```
Don't need `index` here if it's not going to be used.

ryguyk reviewed (commented):
This looks very good overall.  What might seem like a lot of comments below does not speak against the overall quality.  Just some things I want to discuss.

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+// shuffle array using Fisher-Yates shuffle
+const randomizeOrder = imageArray => {
+  const tempArray = imageArray;
+  let currentIndex = Object.keys(tempArray).length;
```
Shouldn't need `Object.keys...` here. `tempArray.length` should suffice.  If we want to expand this function to be usable by more than just

allisonlu commented on src/applications/static-pages/homepage-hero-randomizer/sass/homepage-hero.scss:
```diff
@@ -0,0 +1,48 @@
+.homepage-hero {
+  // gradient background from left to right
+  &__wrapper {
+    background: linear-gradient(to right, $color-old-browser-background-start 37%, $color-old-browser-background-end);
```
Mostly guesstimating since the inspector on Sketch isn't precise. This screenshot was taken from [RT's Sketch file](https://www.sketch.com/s/3aa40506-4be2-46cc-876b-93f1a9f3a857/a/v8Ag388#Inspector). Notice how there is a slight linear gradient from left to right across the whole background color of the hero (you can also see it by inspecting it). 

When setting leaving it as default (0% to 100%), the change in gradient color occurs at 50%, but it seems like the color change is happening more to the left.

<img width="1292" alt="Screen Shot 2022-04-21 at 2 20 54 PM" src="https://user-images.githubusercontent.com/8542413/164529865-9e7aaf26-10f7-4ae6-80d8-868514ff6b34.png">

allisonlu committed:
VACMS-8360:
- fixed eslint errors
- refactored imagePaths array

allisonlu committed:
VACMS-8360:
- added documentation
- removed Object.keys(...)

allisonlu committed:
Merge branch 'master' into VACMS-8360-home-hero-random-image

ryguyk commented on src/applications/static-pages/homepage-hero-randomizer/index.jsx:
```diff
@@ -0,0 +1,62 @@
+import React, { useEffect, useState } from 'react';
+import './sass/homepage-hero.scss';
```
scss

allisonlu committed:
VACMS-8360:
- moved scss import to file where widget is mounted
- moved documentation block

allisonlu committed:
Merge branch 'feature/VACMS-8340-va.gov-homepage-iteration' into VACMS-8360-home-hero-random-image
